src/main/java/io/datareplication/consumer/StreamingPage.java

```diff
@@ -1,24 +1,230 @@
+        public static class BodyChunk<EntityHeader> extends Chunk<EntityHeader> {
+            /**
+             * The chunk's byte data. The ByteBuffer will have its entire capacity used
+             * (i.e. {@link ByteBuffer#position()} == 0 and {@link ByteBuffer#limit() == capacity}). When returned
+             * from {@link StreamingPage}, the buffer will not be shared and may be freely modified.
+             */
+            @NonNull ByteBuffer data;
+        }
+
+        /**
+         * <p>
+         * Represents the end of an entity. It signals that the current entity's entire body has been read and
+         * returned.
+         * </p>
+         *
+         * <p>
+         * May be followed by:
+         * <ul>
+         *     <li>{@link Header}: begin a new entity</li>
+         *     <li>nothing: after the {@link BodyEnd} for the final entity on the page, the stream will terminate</li>
+         * </ul>
+         * </p>
+         *
+         * @param <EntityHeader>
+         */
+        @Value
+        @AllArgsConstructor(access = AccessLevel.PRIVATE)
+        @EqualsAndHashCode(callSuper = false)
+        public static class BodyEnd<EntityHeader> extends Chunk<EntityHeader> {
+            private static final BodyEnd<?> INSTANCE = new BodyEnd<>();
+        }
 
-    // TODO: error handling
-    // TODO: better reader
-    // TODO: content type and content length are inaccessible
-    @NonNull CompletionStage<@NonNull Optional<@NonNull EntityHeader>> next();
+        /**
+         * Create a new {@link Header}.
+         *
+         * @param header         the entity's header object
+         * @param contentType    the entity's content type
+         * @param <EntityHeader> the entity's header
+         * @return a new {@link Header}
+         */
+        public static @NonNull <EntityHeader> Header<EntityHeader> header(
+            @NonNull EntityHeader header,
+            @NonNull ContentType contentType) {
+            return new Header<>(header, contentType);
+        }
+
+        /**
+         * Create a new {@link BodyChunk}.
+         *
+         * @param data           the data chunk; this buffer <strong>should</strong> be {@link ByteBuffer#slice()}d and not
+         *                       shared to avoid ByteBuffer's mutability leaking out and causing problems
+         * @param <EntityHeader> the entity's header
+         * @return a new {@link BodyChunk}
+         */
+        public static @NonNull <EntityHeader> BodyChunk<EntityHeader> bodyChunk(@NonNull ByteBuffer data) {
+            return new BodyChunk<>(data);
+        }
+
+        /**
+         * Return a {@link BodyEnd} marker.
+         *
+         * @param <EntityHeader> the entity's header
+         * @return a {@link BodyEnd} marker
+         */
+        public static @NonNull <EntityHeader> BodyEnd<EntityHeader> bodyEnd() {
+            //noinspection unchecked
+            return (BodyEnd<EntityHeader>) BodyEnd.INSTANCE;
```

Comment:
Set

`BodyEnd<?> INSTANCE` to `BodyEnd<EntityHeader> INSTANCE`

Should eliminate this cast?!